### PR TITLE
feat(tunes): add methods for interacting with tunes

### PR DESCRIPTION
**Pull request by pilartomas**

### CODE REVIEW

File: src/api-types.ts [hunk 1]
```diff
@@ -200,3 +200,72 @@ export interface ModelOutput {
...
+
+export const TuneInputSchema = z.object({
+  name: z.string(),
+  model_id: z.string(),
+  task_id: z.string().nullish(),
+  training_file_ids: z.array(z.string()),
+  validation_file_ids: z.array(z.string()).nullish(),
+  evaluation_file_ids: z.array(z.string()).nullish(),
+  method_id: z.string(),
+  parameters: z.object({}).nullish(),
```
Comment (Tomas2D) [inline src/api-types.ts:240]: `parameters: z.object({}). passthrough().nullish()`
https://zod.dev/?id=passthrough

File: src/api-types.ts [hunk 2]
```diff
@@ -200,3 +200,72 @@ export interface ModelOutput {
...
+});
+export type TuneInput = z.infer<typeof TuneInputSchema>;
```
Comment (Tomas2D) [inline src/api-types.ts:242]: `export type TuneInput = z.input<typeof TuneInputSchema>;`

File: src/client.ts [hunk 3]
```diff
@@ -48,6 +57,7 @@ import { lookupApiKey, lookupEndpoint } from './helpers/config.js';
 import { RETRY_ATTEMPTS_DEFAULT } from './constants.js';
 import { EventStreamContentType, fetchEventSource } from './utils/sse/sse.js';
 import { Transform, TransformCallback } from 'stream';
+import { Readable, Stream } from 'node:stream';
```
Comment (Tomas2D) [inline src/client.ts:60]: `Readable` is unused

File: src/client.ts [hunk 4]
```diff
@@ -695,4 +705,164 @@ export class Client {
...
+        optionsOrCallback,
+        callback,
+      },
+      async ({ input, options }) => {
+        const params = new URLSearchParams();
+        if (input?.filters?.search) params.set('search', input.filters.search);
+        if (input?.filters?.status) params.set('status', input.filters.status);
+
+        let offset = 0;
+        let allCount = Infinity;
```
Comment (Tomas2D) [inline src/client.ts:739]: `totalCount` fits better 👍🏻

File: src/client.ts [hunk 5]
```diff
@@ -695,4 +705,164 @@ export class Client {
...
+          ApiTypes.TuneOutput,
+          ApiTypes.TuneInput
+        >({
+          ...options,
+          method: 'POST',
+          url: `/v1/tunes`,
+          data: input,
+        });
+      }
+      const status = apiOutput.results.status;
```
Comment (Tomas2D) [inline src/client.ts:812]: `const { status } = apiOutput.results;`

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: src/api-types.ts [hunk 6]
```diff
@@ -200,3 +200,72 @@ export interface ModelOutput {
...
+  'PENDING',
+  'HALTED',
+  'RUNNING',
+  'QUEUED',
+  'COMPLETED',
+  'FAILED',
+]);
+export type TuneStatus = z.infer<typeof TuneStatusSchema>;
+
+type TuneFile = {
```
Comment (Tomas2D) [inline src/api-types.ts:218]: use `interface`

File: src/client.ts [hunk 7]
```diff
@@ -593,6 +600,11 @@ export class Client {
           url: '/v1/generate/config',
           stream: false,
           data: input,
+          cache: {
+            update: {
+              [GET_GENERATE_CONFIG_ID]: 'delete',
+            },
+          },
```
Comment (pilartomas) [inline src/client.ts:607]: bugfix

File: src/client.ts [hunk 8]
```diff
@@ -695,4 +707,172 @@ export class Client {
...
+    >(
+      {
+        inputOrOptionsOrCallback: inputOrCallback,
+        optionsOrCallback,
+        callback,
+      },
+      ({ input, options }) => {
+        const params = new URLSearchParams();
+        if (input?.filters?.search) params.set('search', input.filters.search);
+        if (input?.filters?.status) params.set('status', input.filters.status);
```
Comment (Tomas2D) [inline src/client.ts:741]: This part is not used, but it should be.

File: src/helpers/common.ts [hunk 9]
```diff
@@ -105,3 +135,56 @@ export function isNullish<T>(
...
+        for await (const result of generator) {
+          callback(null, result);
+        }
+      } catch (err) {
+        callback(err);
+      }
+    })();
+    return;
+  };
+}
```
Comment (Tomas2D) [inline src/helpers/common.ts:153]: Why not just this?

```typescript
export function callbackifyGenerator<T>(generatorFn: () => AsyncGenerator<T>) {
  return async (callback: AnyFn) => {
      try {
        for await (const result of generatorFn()) {
          callback(null, result);
        }
      } catch (err) {
        callback(err);
      }
  };
}
```

or

```typescript
export function callbackifyGenerator<T>(generatorFn: () => AsyncGenerator<T>) {
  return (callback: AnyFn) => {
      (async () => {
        try {
          for await (const result of generatorFn()) {
            callback(null, result);
          }
        } catch (err) {
          callback(err);
        }
      })();
  };
}
```

File: src/client.ts [hunk 10]
```diff
@@ -695,4 +707,172 @@ export class Client {
...
+        return paginator(
+          (paginatorParams) =>
+            this.#fetcher<ApiTypes.TunesOuput>({
+              ...options,
```
Comment (Tomas2D) [inline src/client.ts:745]: probably here the `params` variable should be passed

File: src/client.ts [hunk 11]
```diff
@@ -695,4 +707,172 @@ export class Client {
...
+          cache: {
+            update: {
+              [GET_TUNE_ID]: 'delete',
+            },
+          },
+        });
+      }
+      const { status } = apiOutput.results;
+      switch (status) {
+        case 'COMPLETED':
```
Comment (Tomas2D) [inline src/client.ts:824]: Just a simple `if` statement would be better, but I don't dwell on it.

File: src/helpers/common.ts [hunk 12]
```diff
@@ -105,3 +135,56 @@ export function isNullish<T>(
...
+    );
+    const output = await executor(paginatedSearchParams);
+    for (const result of output.results) {
+      yield result;
+      if (--remainingCount === 0) return;
+      ++currentOffset;
+    }
+    totalCount = output.totalCount;
+  }
+}
```
Comment (Tomas2D) [inline src/helpers/common.ts:190]: Can we give the client information about the total amount, as we have this information from the API?

File: src/tests/integration/client.test.ts [hunk 13]
```diff
@@ -119,4 +121,71 @@ describe('client', () => {
...
+    test('should download assets of a completed tune', async () => {
+      const {
+        id,
+        assets: { encoder },
+      } = tunesStore[0];
+      const tune = await client.tune({ id });
+      expect.assertions(1);
+      if (tune.status === 'COMPLETED') {
+        const content = await tune.downloadAsset('encoder');
+        // unfortunately, we don't have .toArray() in the supported version of Node
```
Comment (Tomas2D) [inline src/tests/integration/client.test.ts:168]: Soon we will have [fromAsync](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fromAsync) 👍🏻

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: src/tests/integration/client.test.ts (hunk 13 above)
Comment (pilartomas) [reply src/tests/integration/client.test.ts:168]: yeah sorry, `fromAsync` 👍

**Review by Tomas2D — APPROVED** [review]

**Pull request merged**